File: scripts/qwenimage_encode.py
#!/usr/bin/env python
"""Qwen-Image-2512 — Stage 1 (text encode).

Loads Qwen2.5-VL-7B from the Qwen-Image-2512 repo, applies the exact
diffusers prompt template, drops the first 34 tokens (system prompt),
extracts only the non-padded hidden states, and saves cond + uncond
embeddings to a safetensors file. Never loads the DiT or VAE.

Usage:
    python qwenimage_encode.py \\
        "a photograph of an astronaut riding a horse on mars" \\
        ""  \\
        /path/to/embeds.safetensors

The saved file contains two BF16 tensors (shapes match the DROPPED,
non-padded output — they will typically have different seq lengths):
    cond:   [1, L_cond,   3584]
    uncond: [1, L_uncond, 3584]

Stage 2 (Rust) loads these and runs the DiT denoise loop.
Stage 3 (Python) runs the VAE decode.
"""
from __future__ import annotations

import os
import sys
import time
from pathlib import Path

import torch
from safetensors.torch import save_file
from transformers import AutoTokenizer, Qwen2_5_VLForConditionalGeneration

# Diffusers pipeline constants (pipeline_qwenimage.py:175-177)
PROMPT_TEMPLATE_ENCODE = (
    "<|im_start|>system\nDescribe the image by detailing the color, shape, "
    "size, texture, quantity, text, spatial relationships of the objects "
    "and background:<|im_end|>\n<|im_start|>user\n{}<|im_end|>\n"
    "<|im_start|>assistant\n"
)
PROMPT_TEMPLATE_ENCODE_START_IDX = 34
TOKENIZER_MAX_LENGTH = 1024

REPO = "Qwen/Qwen-Image-2512"
SNAP = (
    "/home/alex/.cache/huggingface/hub/models--Qwen--Qwen-Image-2512/snapshots/"
    "25468b98e3276ca6700de15c6628e51b7de54a26"
)


def extract_masked_hidden(hidden_states: torch.Tensor, mask: torch.Tensor):
    """Return the non-padded hidden states for each sample (diffusers
    _extract_masked_hidden). For single-prompt input this returns one
    tensor of shape [valid_len, hidden]."""
    bool_mask = mask.bool()
    valid_lengths = bool_mask.sum(dim=1)
    selected = hidden_states[bool_mask]
    return torch.split(selected, valid_lengths.tolist(), dim=0)


def encode_prompt(
    tokenizer,
    text_encoder,
    prompt: str,
    device: str,
    dtype: torch.dtype,
) -> torch.Tensor:
    """Mirror of `_get_qwen_prompt_embeds` → drop system-prompt tokens,
    return the [1, L, hidden] tensor for a single prompt."""
    txt = PROMPT_TEMPLATE_ENCODE.format(prompt)
    drop_idx = PROMPT_TEMPLATE_ENCODE_START_IDX

    txt_tokens = tokenizer(
        [txt],
        max_length=TOKENIZER_MAX_LENGTH + drop_idx,
        padding=True,
        truncation=True,
        return_tensors="pt",
    ).to(device)

    with torch.no_grad():
        out = text_encoder(
            input_ids=txt_tokens.input_ids,
            attention_mask=txt_tokens.attention_mask,
            output_hidden_states=True,
        )
    hidden_states = out.hidden_states[-1]
    split = extract_masked_hidden(hidden_states, txt_tokens.attention_mask)
    split = [e[drop_idx:] for e in split]  # drop the 34 system-prompt tokens

    # For a single prompt, return [1, L, H]
    emb = split[0].unsqueeze(0).to(dtype=dtype, device=device)
    return emb


def main() -> int:
    prompt = sys.argv[1] if len(sys.argv) > 1 else (
        "a photograph of an astronaut riding a horse on mars, "
        "cinematic lighting, highly detailed"
    )
    negative = sys.argv[2] if len(sys.argv) > 2 else ""
    out_path = sys.argv[3] if len(sys.argv) > 3 else (
        "/home/alex/serenity/output/qwenimage_embeds.safetensors"
    )

    device = "cuda"
    dtype = torch.bfloat16

    print("=== Qwen-Image-2512 — Stage 1 (text encode) ===")
    print(f"Prompt:   {prompt!r}")
    print(f"Negative: {negative!r}")
    print(f"Output:   {out_path}")
    print()

    snap = Path(SNAP)
    if not snap.exists():
        print(f"[error] snapshot path does not exist: {snap}")
        print("       download the text_encoder shards first via:")
        print(f"       hf_hub_download('{REPO}', 'text_encoder/...')")
        return 2

    print("--- Loading tokenizer ---")
    tokenizer = AutoTokenizer.from_pretrained(snap / "tokenizer")

    print("--- Loading Qwen2.5-VL-7B text encoder ---")
    t0 = time.time()
    text_encoder = Qwen2_5_VLForConditionalGeneration.from_pretrained(
        snap / "text_encoder",
        dtype=dtype,
        device_map=device,
    )
    text_encoder.eval()
    print(f"  Loaded in {time.time() - t0:.1f}s")
    print()

    print("--- Encoding cond ---")
    t0 = time.time()
    cond = encode_prompt(tokenizer, text_encoder, prompt, device, dtype)
    print(f"  cond:   {tuple(cond.shape)} in {time.time() - t0:.1f}s")

    print("--- Encoding uncond ---")
    t0 = time.time()
    uncond = encode_prompt(tokenizer, text_encoder, negative, device, dtype)
    print(f"  uncond: {tuple(uncond.shape)} in {time.time() - t0:.1f}s")

    # Free VRAM
    del text_encoder
    torch.cuda.empty_cache()
    print("  Text encoder evicted")
    print()

    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    tensors = {
        "cond": cond.contiguous().cpu(),
        "uncond": uncond.contiguous().cpu(),
    }
    save_file(tensors, out_path)
    print(f"EMBEDDINGS SAVED: {out_path}")
    print(f"  cond:   {tuple(tensors['cond'].shape)}  {tensors['cond'].dtype}")
    print(f"  uncond: {tuple(tensors['uncond'].shape)}  {tensors['uncond'].dtype}")
    print()
    print(f"Next: ./target/release/qwenimage_gen {out_path} <latents.safetensors>")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/qwenimage_encode.rs
//! Qwen-Image-2512 — Stage 1 (Rust text encode).
//!
//! Port of `scripts/qwenimage_encode.py`. Loads the Qwen2.5-VL-7B text
//! encoder (4 safetensors shards) through the in-tree `Qwen25VLEncoder`,
//! applies the diffusers prompt template, drops the first 34 system-
//! prompt tokens, trims to non-padded length, and writes cond + uncond
//! embeddings to a safetensors file in the exact format `qwenimage_gen`
//! expects.
//!
//! Usage:
//!     target/release/qwenimage_encode \
//!         "a photograph of an astronaut riding a horse on mars" \
//!         ""  \
//!         /path/to/embeds.safetensors
//!
//! Knobs:
//!     QWEN_TE_MAX_LEN   — tokenizer max length (default 1024).
//!     QWEN_TE_SHARDS    — colon-separated shard paths (default: Qwen-Image-2512 snapshot).
//!     QWEN_TOKENIZER    — tokenizer.json path (default: Qwen-Image-2512 snapshot).
//!
//! Stage 2 then consumes the written embeddings:
//!     target/release/qwenimage_gen <embeds.safetensors> <out.png>

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::time::Instant;

use flame_core::{global_cuda_device, DType, Shape, Tensor};

use inference_flame::models::qwen25vl_encoder::{Qwen25VLConfig, Qwen25VLEncoder};

const SNAP: &str = "/home/alex/.serenity/models/checkpoints/qwen-image-2512";

/// System prefix + user wrapper applied before tokenization.
/// Matches `pipeline_qwenimage.py::PROMPT_TEMPLATE_ENCODE`.
const PROMPT_PREFIX: &str =
    "<|im_start|>system\nDescribe the image by detailing the color, shape, size, \
     texture, quantity, text, spatial relationships of the objects and background:\
     <|im_end|>\n<|im_start|>user\n";
const PROMPT_SUFFIX: &str = "<|im_end|>\n<|im_start|>assistant\n";

/// Number of leading tokens to drop (the system prompt). Match Python
/// constant `PROMPT_TEMPLATE_ENCODE_START_IDX`.
const DROP_IDX: usize = 34;

/// Qwen2.5-VL pad token id.
const PAD_TOKEN: i32 = 151643;

fn text_encoder_shards() -> Vec<PathBuf> {
    if let Ok(s) = std::env::var("QWEN_TE_SHARDS") {
        return s.split(':').map(PathBuf::from).collect();
    }
    (1..=4)
        .map(|i| PathBuf::from(format!("{SNAP}/text_encoder/model-0000{i}-of-00004.safetensors")))
        .collect()
}

fn tokenizer_path() -> PathBuf {
    if let Ok(p) = std::env::var("QWEN_TOKENIZER") {
        return PathBuf::from(p);
    }
    PathBuf::from(format!("{SNAP}/tokenizer/tokenizer.json"))
}

fn env_usize(key: &str, default: usize) -> usize {
    std::env::var(key)
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(default)
}

/// Tokenize with the Qwen template, pad to `max_len`, truncate if over.
fn tokenize_prompt(
    tokenizer: &tokenizers::Tokenizer,
    prompt: &str,
    max_len: usize,
) -> anyhow::Result<Vec<i32>> {
    let wrapped = format!("{PROMPT_PREFIX}{prompt}{PROMPT_SUFFIX}");
    let enc = tokenizer
        .encode(wrapped, false)
        .map_err(|e| anyhow::anyhow!("tokenize: {e}"))?;
    let ids = enc.get_ids();
    let mut out: Vec<i32> = ids.iter().take(max_len).map(|&i| i as i32).collect();
    while out.len() < max_len {
        out.push(PAD_TOKEN);
    }
    Ok(out)
}

/// Run the encoder, drop the first `DROP_IDX` tokens, trim trailing
/// pad tokens. Returns `[1, L, 3584]` BF16.
fn encode_and_trim(
    encoder: &Qwen25VLEncoder,
    token_ids: &[i32],
) -> anyhow::Result<Tensor> {
    let hidden = encoder.encode(token_ids)?; // [1, seq_len, 3584]
    let dims = hidden.shape().dims().to_vec();
    let seq_len = dims[1];
    let hidden_size = dims[2];

    let real_len = token_ids
        .iter()
        .position(|&id| id == PAD_TOKEN)
        .unwrap_or(seq_len);
    if real_len <= DROP_IDX {
        anyhow::bail!(
            "prompt produced only {real_len} tokens; expected > {DROP_IDX} after Qwen template"
        );
    }
    let kept_len = real_len - DROP_IDX;
    // narrow(1, DROP_IDX, kept_len) gives [1, kept_len, hidden_size]
    let out = hidden.narrow(1, DROP_IDX, kept_len)?;
    let _ = hidden_size;
    Ok(out)
}

fn main() -> anyhow::Result<()> {
    env_logger::try_init().ok();
    let t_total = Instant::now();
    let device = global_cuda_device();

    let args: Vec<String> = std::env::args().collect();
    let prompt = args.get(1).cloned().unwrap_or_else(|| {
        "a photograph of an astronaut riding a horse on mars, cinematic lighting, highly detailed"
            .to_string()
    });
    let negative = args.get(2).cloned().unwrap_or_default();
    let out_path = args.get(3).cloned().unwrap_or_else(|| {
        "/home/alex/EriDiffusion/inference-flame/output/qwenimage_embeds.safetensors".to_string()
    });

    let max_len = env_usize("QWEN_TE_MAX_LEN", 1024) + DROP_IDX;

    println!("=== Qwen-Image-2512 — Stage 1 (Rust text encode) ===");
    println!("Prompt:   {:?}", prompt);
    println!("Negative: {:?}", negative);
    println!("Output:   {out_path}");
    println!();

    // Tokenizer
    let tok_path = tokenizer_path();
    if !tok_path.exists() {
        anyhow::bail!(
            "tokenizer.json not found at {:?}. Regenerate with: \
             python -c \"from transformers import AutoTokenizer as T; \
             T.from_pretrained('{}/tokenizer').save_pretrained('{}/tokenizer')\"",
            tok_path,
            SNAP,
            SNAP
        );
    }
    let tokenizer = tokenizers::Tokenizer::from_file(&tok_path)
        .map_err(|e| anyhow::anyhow!("load tokenizer from {:?}: {e}", tok_path))?;

    println!("--- Tokenizing ---");
    let pos_ids = tokenize_prompt(&tokenizer, &prompt, max_len)?;
    let neg_ids = tokenize_prompt(&tokenizer, &negative, max_len)?;
    let pos_real = pos_ids.iter().position(|&i| i == PAD_TOKEN).unwrap_or(pos_ids.len());
    let neg_real = neg_ids.iter().position(|&i| i == PAD_TOKEN).unwrap_or(neg_ids.len());
    println!("  pos: {pos_real}/{max_len} real tokens (trimmed output: {})", pos_real - DROP_IDX);
    println!("  neg: {neg_real}/{max_len} real tokens (trimmed output: {})", neg_real - DROP_IDX);
    println!();

    // Load encoder shards (BF16 → BF16). Filter to text-only keys (drop
    // `visual.*` so we don't carry ~7B of vision weights into GPU RAM).
    println!("--- Loading text encoder shards ---");
    let t_load = Instant::now();
    let mut weights: HashMap<String, Tensor> = HashMap::new();
    for shard in text_encoder_shards() {
        let shard_weights = flame_core::serialization::load_file_filtered(
            &shard,
            &device,
            |k| !k.starts_with("visual."),
        )?;
        for (k, v) in shard_weights {
            weights.insert(k, v);
        }
    }
    // Upcast F32/F16 shards to BF16 if needed.
    let keys: Vec<String> = weights.keys().cloned().collect();
    for k in keys {
        let t = &weights[&k];
        if t.dtype() != DType::BF16 {
            let bf = t.to_dtype(DType::BF16)?;
            weights.insert(k, bf);
        }
    }
    println!(
        "  loaded {} tensors in {:.1}s",
        weights.len(),
        t_load.elapsed().as_secs_f32()
    );

    println!("--- Building encoder ---");
    let mut cfg = Qwen25VLEncoder::config_from_weights(&weights)?;
    cfg.max_seq_len = max_len.max(cfg.max_seq_len);
    println!(
        "  config: layers={} hidden={} heads={}/{} head_dim={} max_seq={}",
        cfg.num_layers, cfg.hidden_size, cfg.num_heads, cfg.num_kv_heads, cfg.head_dim, cfg.max_seq_len
    );
    let encoder = Qwen25VLEncoder::new(weights, cfg, device.clone());

    println!();
    println!("--- Encoding ---");
    let t_enc = Instant::now();
    let cond = encode_and_trim(&encoder, &pos_ids)?;
    let uncond = encode_and_trim(&encoder, &neg_ids)?;
    println!(
        "  cond: {:?}  uncond: {:?}  in {:.1}s",
        cond.shape().dims(),
        uncond.shape().dims(),
        t_enc.elapsed().as_secs_f32()
    );

    drop(encoder);

    // Save in the exact format qwenimage_gen.rs expects.
    if let Some(parent) = Path::new(&out_path).parent() {
        std::fs::create_dir_all(parent).ok();
    }
    let mut out = HashMap::new();
    out.insert("cond".to_string(), cond);
    out.insert("uncond".to_string(), uncond);
    flame_core::serialization::save_file(&out, Path::new(&out_path))?;

    let dt = t_total.elapsed().as_secs_f32();
    println!();
    println!("============================================================");
    println!("EMBEDDINGS SAVED: {out_path}");
    println!("Total time:       {dt:.1}s");
    println!("============================================================");
    println!();
    println!("Next: target/release/qwenimage_gen {out_path} <output.png>");

    let _ = device; // keep device alive
    Shape::from_dims(&[1]); // touch an import so unused-warnings are clean
    Ok(())
}
